From 66c0a0b32d8a3477e413706b7fe6254e44eb16b0 Mon Sep 17 00:00:00 2001
From: anlaneg <anlaneg@example.com>
Date: Mon, 29 Jan 2018 16:45:22 +0800
Subject: [PATCH 13/13] remove sudo

---
 Setup.sh | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/Setup.sh b/Setup.sh
index ec7e017..fd441df 100755
--- a/Setup.sh
+++ b/Setup.sh
@@ -1,2 +1,7 @@
 #! /bin/bash
-sudo ./tools/vnf_build.sh -s -d=17.05
+if [ ! `id -u` -eq  0 ];
+then
+    echo "Checking for user permission.. Password-less sudo user must run this script"
+    exit 1
+fi;
+./tools/vnf_build.sh -s -d=17.05
-- 
2.7.4

