--- ./dpdk/drivers/net/nfp/nfp_net.c	2016-11-13 22:28:12.000000000 +0800
+++ ./dpdk/drivers/net/nfp/nfp_net.c	2017-12-14 14:40:03.053585228 +0800
@@ -1801,7 +1801,8 @@
 		nfp_net_rx_cksum(rxq, rxds, mb);
 
 		/* Checking the port flag */
-		nfp_net_check_port(rxds, mb);
+		//nfp_net_check_port(rxds, mb);
+		mb->port = rxq->port_id;
 
 		if ((rxds->rxd.flags & PCIE_DESC_RX_VLAN) &&
 		    (hw->ctrl & NFP_NET_CFG_CTRL_RXVLAN)) {
